### ISSUE TITLE
Internal Sharding

### ISSUE THREAD

**Issue by bdistin** [hydrabolt/discord.js#1414]
Starting with v12.0.0 add an option for internal sharding. This would be different than the traditional sharding manager, in that your one Discord Client/process would handle sharding with discord rather than spawning multiple processes with their own Discord Client instances each.

Benefits:

- Memory for small/mid sized bots would be less than the sum of traditional sharding shards due to the following reasons:
-- The base instance of your bot with all dependencies before anything is cached, is not duplicated for each shard.
-- Users which are shared with guilds of different shards aren't duplicated across caches (because it is one collection for users, instead of one for each shard).
- It is easier and more efficient to manage, in that to perform simple tasks you do not have to broadcast eval to all shards, and do the computations to eliminate duplications from your results. For instance, getting the number of unique users your bot has across all shards.
- It is easier to implement for the bot dev. Work In Progress implementations/PRs from @GusCaplan has been as easy as setting the client option `shardCount` to an integer (number of shards) or as the string 'auto' to automatically determine how many internal shards should be created.
- Still has the benefit of Traditional sharding over not sharding, in which if a websocket disconnects or receives a 1000 code and cannot reconnect or resume, the other shards remain up and unaffected by the outage.

Drawbacks:

- Not viable for large/very large bots due to the single threaded nature of JavaScript. Such bots should use traditional sharding.

*Sizes in this issue are based on the frequency of events handled, not via number of guilds.*

**Comment by hydrabolt** [comment]
This seems good, shouldn't be too hard to implement once the WebSocket rewrite is merged into master as it's mainly just a case of creating more `WebSocketConnection`s 👍

**Comment by CakeNetwork** [comment]
So will bots be auto-shardingnow or will the option to turn on / off internal sharding be a thing?

**Comment by JacobGunther12** [comment]
+1

**Comment by devsnek** [comment]
this is being handled by #1501

**Comment by PassTheMayo** [comment]
@DiscordDonut That would probably be true. I don't see any reason why it would be turned off by normal.

**Comment by Gawdl3y** [comment]
It would be turned off by default because it's dumb.

**Comment by Kashalls** [comment]
Is this still in progress?

**Comment by Dev-Yukine** [comment]
look at the Internal Sharding branch, this branch is currently in testing of some bots from D.js staff and after they are sure that it works correct without bugs they will merge it into master.
